geopandas/io/tests/test_infer_schema.py

```diff
@@ -0,0 +1,227 @@
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': 'Polygon', 'properties': OrderedDict()}))
+
+    def test_infer_schema_when_dataframe_has_polygons_and_multipolygons(self):
+        df = GeoDataFrame(
+            {},
+            crs={'init': 'epsg:4326', 'no_defs': True},
+            geometry=[MultiPolygon((self.city_hall_boundaries, self.vauquelin_place)), self.city_hall_boundaries]
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': contains_inanyorder('MultiPolygon', 'Polygon'), 'properties': OrderedDict()}))
+
+    def test_infer_schema_when_dataframe_has_only_multipolygons(self):
+        df = GeoDataFrame(
+            {},
+            crs={'init': 'epsg:4326', 'no_defs': True},
+            geometry=[MultiPolygon((self.city_hall_boundaries, self.vauquelin_place))]
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': 'MultiPolygon', 'properties': OrderedDict()}))
+
+    def test_infer_schema_when_dataframe_has_multiple_shape_types(self):
+        df = GeoDataFrame(
+            {},
+            crs={'init': 'epsg:4326', 'no_defs': True},
+            geometry=[MultiPolygon((self.city_hall_boundaries, self.vauquelin_place)),
+                      self.city_hall_boundaries,
+                      MultiLineString(self.city_hall_walls),
+                      self.city_hall_walls[0],
+                      MultiPoint([self.city_hall_entrance, self.city_hall_balcony]),
+                      self.city_hall_balcony]
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': contains_inanyorder('MultiPolygon', 'Polygon', 'MultiLineString', 'LineString', 'MultiPoint', 'Point'),
+                                                   'properties': OrderedDict()}))
+
+    def test_infer_schema_when_dataframe_has_a_3D_shape_type(self):
+        df = GeoDataFrame(
+            {},
+            crs={'init': 'epsg:4326', 'no_defs': True},
+            geometry=[MultiPolygon((self.city_hall_boundaries, self.vauquelin_place)),
+                      self.city_hall_boundaries,
+                      MultiLineString(self.city_hall_walls),
+                      self.city_hall_walls[0],
+                      MultiPoint([self.city_hall_entrance, self.city_hall_balcony]),
+                      self.city_hall_balcony,
+                      self.point_3D]
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': contains_inanyorder('3D MultiPolygon', '3D Polygon', '3D MultiLineString', '3D LineString', '3D MultiPoint', '3D Point'), 'properties': OrderedDict()}))
+
+    def test_infer_schema_when_dataframe_has_a_3D_Point(self):
+        df = GeoDataFrame(
+            {},
+            crs={'init': 'epsg:4326', 'no_defs': True},
+            geometry=[self.city_hall_balcony, self.point_3D]
+        )
+
+        assert_that(infer_schema(df), has_entries({'geometry': '3D Point', 'properties': OrderedDict()}))
```

Comment:
I am wondering for this case: do we only need to specify "3D Point", or should we have both `('3D Point', 'Point')`? 

(I don't really remember how fiona is handling the 3D things)

Reply:
I agree: inferred schema would be more accurate.

I'll try it, and see how Fiona is reacting

Reply:
@jorisvandenbossche: this change makes the build fail; the result can be found in [build 1931](https://travis-ci.org/geopandas/geopandas/builds/466303881)

Here is what I experienced when mixing 3D and 2D shapes:

- when writing with `GeoJSON` driver, all tests pass with for **all** current build configurations
- when writing with `ESRI Shapefile` driver, tests where3D and 2D shapes are mixed together fail for **some** build configurations (probably the ones with Fiona older releases, but I'm not sure)

maybe you will be able to explain the differences between drivers

For now I'm trying to implement 2 different behaviors to `infer_schema` function, depending on Fiona version. I plan using some code introduced recently:

```
_FIONA18 = LooseVersion(fiona.__version__) >= LooseVersion('1.8')
```
let's keep in touch!

Reply:
@jorisvandenbossche: this "_before/after Fiona 1.8_" strategy fixed all existing unit tests. Is it an appropriate coding scheme?
If it is, I would like to add some other changes, because there are still cases (using `ERSI Shapefile` driver) where `to_file` function fails.